src/ipred.rs

```diff
@@ -979,33 +993,28 @@ unsafe fn ipred_z2_rust<BD: BitDepth>(
         };
         if filter_strength_0 != 0 {
             filter_edge::<BD>(
-                &mut *topleft.offset(-height as isize),
+                &mut edge,
+                topleft - height as usize,
                 height,
                 height - max_height,
                 height,
-                &*topleft_in.offset(-height as isize),
+                topleft_in,
+                topleft_in_off - height as usize,
                 0 as c_int,
                 height + 1,
                 filter_strength_0,
             );
         } else {
             BD::pixel_copy(
-                slice::from_raw_parts_mut(
-                    topleft.offset(-height as isize),
-                    height.try_into().unwrap(),
-                ),
-                slice::from_raw_parts(
-                    topleft_in.offset(-height as isize),
-                    height.try_into().unwrap(),
-                ),
+                &mut edge[topleft - height as usize..][..height as usize],
+                &topleft_in[topleft_in_off - height as usize..][..height as usize],
                 height.try_into().unwrap(),
             );
         }
     }
-    *topleft = *topleft_in;
+    edge[topleft] = topleft_in[topleft_in_off];
     let base_inc_x = 1 + upsample_above as c_int;
-    let left: *const BD::Pixel =
-        &mut *topleft.offset(-(1 + upsample_left as isize)) as *mut BD::Pixel;
+    let left = edge[topleft - (1 + upsample_left as usize)..].as_mut_ptr();
```

Comment:
I would try to make `left` safe as part of this PR, as it could reveal borrowck issues that influence the design of the rest of this PR.